### PR TITLE
Porting SDK to Direct 3.0.0.25-preview

### PR DESCRIPTION
**Pull request by kirankumarkolli**
Porting SDK to Direct 3.0.0.25-preview

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/ClientRetryPolicy.cs [hunk 1]
```diff
@@ -119,42 +141,41 @@ public void OnBeforeSendRequest(DocumentServiceRequest request)
...
-            SubStatusCodes? subStatusCode,
-            Func<Task<ShouldRetryResult>> throttlingRetry)
+            SubStatusCodes? subStatusCode)
         {
             if (!statusCode.HasValue
                 && (!subStatusCode.HasValue
                 || subStatusCode.Value == SubStatusCodes.Unknown))
             {
-                return throttlingRetry();
+                return null;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/ClientRetryPolicy.cs]: Won't this cause null reference exceptions?

File: Microsoft.Azure.Cosmos/src/Query/DefaultDocumentQueryExecutionContext.cs [hunk 2]
```diff
@@ -129,92 +126,131 @@ protected override async Task<FeedResponse<CosmosElement>> ExecuteInternalAsync(
...
             // The code leaves some temporary garbage in request (in RequestContext etc.),
             // which shold be erased during retries.
             using (DocumentServiceRequest request = await this.CreateRequestAsync())
             {
-                if (!string.IsNullOrEmpty(request.Headers[HttpConstants.HttpHeaders.PartitionKey])
-                    || !request.ResourceType.IsPartitioned())
+                FeedResponse<CosmosElement> feedRespose;
+                string partitionIdentifier;
+                // We need to determine how to execute the request:
+                if (LogicalPartitionKeyProvided(request))
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/DefaultDocumentQueryExecutionContext.cs]: A lot of this code was refactored and forked for the v3 to convert to items. We will probably need to port the changes to CosmosGatewayQueryExecutionContext.

File: Microsoft.Azure.Cosmos/src/ClientRetryPolicy.cs (hunk 1 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/ClientRetryPolicy.cs]: Its an internal state not the end state.

File: Microsoft.Azure.Cosmos/src/Query/DefaultDocumentQueryExecutionContext.cs (hunk 2 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Query/DefaultDocumentQueryExecutionContext.cs]: Yes, its a follow-up item. Lets also close with Brandon on path forward.

**Comment by kirankumarkolli** [comment]
@bchong95 can you please take a look at query parts.

**Pull request merged**